src/game/game_init.c

```diff
@@ -683,34 +676,53 @@ void init_controllers(void) {
     gSramProbe = nuPiInitSram();
 #endif
 
-    // Loop over the 4 ports and link the controller structs to the appropriate
-    // status and pad. Interestingly, although there are pointers to 3 controllers,
-    // only 2 are connected here. The third seems to have been reserved for debug
-    // purposes and was never connected in the retail ROM, thus gPlayer3Controller
-    // cannot be used, despite being referenced in various code.
-    for (cont = 0, port = 0; port < 4 && cont < 2; port++) {
+    // Loop over the 4 ports and link the controller structs to the appropriate status and pad.
+    for (cont = 0, port = 0, lastUsedPort = -1; port < MAXCONTROLLERS && cont < NUM_SUPPORTED_CONTROLLERS; port++) {
         // Is controller plugged in?
         if (gControllerBits & (1 << port)) {
+            controller = &gControllers[cont];
+
             // The game allows you to have just 1 controller plugged
             // into any port in order to play the game. this was probably
             // so if any of the ports didn't work, you can have controllers
             // plugged into any of them and it will work.
-#if ENABLE_RUMBLE
-            gControllers[cont].port = port;
-#endif
-            gControllers[cont].statusData = &gControllerStatuses[port];
-            gControllers[cont++].controllerData = &gControllerPads[port];
+            controller->statusData = &gControllerStatuses[port];
+            controller->controllerData = &gControllerPads[port];
+            controller->port = port;
+
+            if (controller->statusData->type & CONT_GCN) {
+                osSyncPrintf("GameCube controller ");
+            } else {
+                osSyncPrintf("N64 controller ");
+            }
+            if (controller->statusData->status & CONT_CARD_ON) {
+                osSyncPrintf("with pak ");
+            }
+            osSyncPrintf("found in port %d\n", port);
```

Comment:
this is great, but a thought for later, i think we should have some prints under different log levels